File: frontend/app/charts/schema.ts
import { z } from "zod";

export interface Fruits{
  /**
   * A UUID for the order.
   */
  name: string;
  /**
   * The name of the product purchased.
   */
  form: string;
  /**
   * The amount of the order.
   */
  retailPrice: number;
  
}

export const filterSchema = (names: string[]) => {
  return z
    .object({
      names: z
        .array(
          z.enum([
            names[0],
            ...names.slice(1, names.length),
          ]),
        )
        .optional()
        .describe(
          `Filter orders by the first name. Lowercase only. MUST only be a list of the following products: ${productNamesAsString}`,
        ),
      retailPrice: z
        .number()
        .min(0)
        .max(100)
        .optional()
        .describe(
          "Retail Price of the Fruit",
        ),
      form: z
        .enum([
          "canned",
          "dried",
          "fresh",
          "frozen",
          "juice",
        ])
        .optional()
        .describe("The form of the fuirt."),
    })
    .describe("Available filters to apply to orders.");
};

export interface Filter {
  name?: string;
  form?: string;
  retailPrice?: string;
  
}

File: backend/gen_ui_backend/charts/schema.py
from typing import List, Literal, Optional, Type

from langchain_core.pydantic_v1 import BaseModel, Field

ChartType = Literal["bar", "line", "pie",]



class Fruits(BaseModel):
    """CamelCase is used here to match the schema used in the frontend."""
    name: str = Field(..., description="Name of fruit")
    form: str = Field(..., description="Form of fruit")
    retailPrice: float = Field(..., description="Retail price of fruit")



class Filter(BaseModel):
    name: Optional[List[str]] = Field(
        None, description="List of product names to filter by"
    )
    form: Optional[str] = Field(
        None, description="Filter orders by form of fruit."
    )
    retailPrice: Optional[float] = Field(
        None, description="Filter orders by retail price"
    )
    


def filter_schema(product_names: List[str]) -> Type[BaseModel]:
    # product_names_as_string = ", ".join(name.lower() for name in product_names)

    class FilterSchema(BaseModel):
        """Available filters to apply to orders."""
        name: Optional[List[str]] = Field(
        None, description="List of product names to filter by"
        )
        form: Optional[str] = Field(
            None, description="Filter orders by form of fruit."
        )
        retailPrice: Optional[float] = Field(
            None, description="Filter orders by retail price"
        )
    

    return FilterSchema


class DataDisplayTypeAndDescription(BaseModel):
    title: str = Field(..., description="The title of the data display type.")
    chartType: ChartType = Field(
        ..., description="The type of chart which this format can be displayed on."
    )
    description: str = Field(
        ..., description="The description of the data display type."
    )
    key: str = Field(..., description="The key of the data display type.")

    class Config:
        allow_population_by_field_name = True
